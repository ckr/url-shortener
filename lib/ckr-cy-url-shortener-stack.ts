import * as cdk from '@aws-cdk/core';

export class CkrCyUrlShortenerStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // The code that defines your stack goes here
  }
}
