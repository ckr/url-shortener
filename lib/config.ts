import { App } from '@aws-cdk/core';
import { ICertificate } from '@aws-cdk/aws-certificatemanager';
import { IHostedZone } from '@aws-cdk/aws-route53';

export class Config {
    private app: App;
    public region: string;
    public urlSuffix: string;
    public zoneName: string;
    public subjectAlternativeNames: string[];
    public zone: IHostedZone;
    public certificate: ICertificate;

    constructor(app: App) {
        this.app = app;
    }
    
    public getApp() {
        return this.app;
    }
    
    public getParam(keyName: string) {
        return this.app.node.tryGetContext(keyName);
    }

    public getStackParam(keyName: string) {
        return this.app.node.tryGetContext('stack_params')[keyName];
    }
}