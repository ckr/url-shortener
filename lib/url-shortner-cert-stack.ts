import { Stack, Construct, StackProps, } from '@aws-cdk/core';
import { Certificate, ICertificate, DnsValidatedCertificate, ValidationMethod, } from '@aws-cdk/aws-certificatemanager';
import { HostedZone, IHostedZone, } from '@aws-cdk/aws-route53';

export class CertificateAttributes {
    zoneName: string;
    subjectAlternativeNames: string[];
    zone: IHostedZone;
    certificate: ICertificate;
}

export class UrlShortenerCertStack extends Stack {
    private certAttr: CertificateAttributes;

    getStackStringParam(key: string): string {
        return this.node.tryGetContext('stack_params')[key];
    }

    getStackListParam(key: string): string[] {
        return this.node.tryGetContext('stack_params')[key] || [];
    }

    constructor(scope: Construct, id: string, certAttr: CertificateAttributes, props?: StackProps) {
        super(scope, id, props);
        this.certAttr = certAttr;

        this.certAttr.zoneName = this.getStackStringParam('dnsZoneName');
        this.certAttr.subjectAlternativeNames = this.getStackListParam('subjectAlternativeNames');

        this.certAttr.zone = HostedZone.fromLookup(this, 'ZoneCK', { domainName: this.certAttr.zoneName });

        this.certAttr.certificate = new DnsValidatedCertificate(this, 'Certificate', {
            domainName: this.certAttr.zoneName,
            subjectAlternativeNames: this.certAttr.subjectAlternativeNames,
            hostedZone: this.certAttr.zone,
            validationMethod: ValidationMethod.DNS,
            region: 'us-east-1' // CloudFront only supports us-east-1 certs from ACM,,,,,,,,,,,
        });
    }
}
