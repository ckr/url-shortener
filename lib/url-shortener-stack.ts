import * as cdk from '@aws-cdk/core';
import { Bucket } from '@aws-cdk/aws-s3';
import { BucketDeployment, Source } from '@aws-cdk/aws-s3-deployment';
import { Duration, RemovalPolicy, Stack, CfnOutput } from '@aws-cdk/core';
import {
    CloudFrontWebDistribution,
    SSLMethod,
    SecurityPolicyProtocol,
    OriginProtocolPolicy,
    CloudFrontAllowedMethods,
    CloudFrontAllowedCachedMethods
} from '@aws-cdk/aws-cloudfront';
import { ARecord, RecordTarget } from '@aws-cdk/aws-route53';
import { CloudFrontTarget } from '@aws-cdk/aws-route53-targets';
import {
    RestApi,
    LambdaIntegration,
    PassthroughBehavior
} from '@aws-cdk/aws-apigateway';
import {
    Function as LamdaFunction,
    Runtime as LamdaRuntime,
    Code as LamdaCode
} from '@aws-cdk/aws-lambda';
import { Config } from './config';

export class UrlShortenerStack extends cdk.Stack {
    constructor(id: string, config: Config) {
        super(config.getApp(), id, {
            env: config.getParam('aws_env_details'),
            tags: config.getParam('stack_tags'),
            description: 'Serverless private URL shortener based on Amazon S3, AWS Lambda, Amazon CloudFront and API Gateway.',
        });

        config.urlSuffix = this.urlSuffix;
        config.region = this.region;
        new UrlShortener(this, 'UrlShortener', config);
    }
}

export class UrlShortener extends cdk.Construct {
    private bucket: Bucket;
    private meRedirect: string;

    /**
     * Creates an s3 bucket with default life cycle rules under the u prefix
     * Configures default index.html and error.html
     * Enables public access
     * Destroyes the bucket if the stack is destroyed
     * Adds an admin folder
     * 
     * @param name The name of the bucket in the cloudformation stack
     * @param bucketName The name of the bucket in s3, has to be unique
     */
    createS3Bucket(name: string, bucketName: string) { // Create a bucket to store the short url files
        this.bucket = new Bucket(this, name, {
            bucketName: bucketName,
            removalPolicy: RemovalPolicy.DESTROY,
            websiteIndexDocument: 'index.html',
            websiteErrorDocument: 'error.html',
            publicReadAccess: true,
            lifecycleRules: [
                {
                    id: 'DisposeShortUrls',
                    prefix: 'u',
                    expiration: Duration.days(7)
                }
            ]
        });
    }

    createPermanentRedirectLink(file: string, redirect: string, distribution: CloudFrontWebDistribution) {
        new BucketDeployment(this, `DeployPLinks-${file}`, {
            sources: [
                Source.asset('./assets/p', {
                    exclude: ['**', `!${file}`]
                })
            ],
            destinationBucket: this.bucket,
            destinationKeyPrefix: 'p',
            contentType: 'text/html',
            websiteRedirectLocation: redirect,
            distribution: distribution,
            distributionPaths: [`/p/${file}`]
        });
    }

    constructor(scope: cdk.Construct, id: string, config: Config) {
        super(scope, id);

        this.createS3Bucket('Bucket', config.zoneName);
        this.meRedirect = config.getStackParam('meRedirect');

        const shortLamda = new LamdaFunction(this, 'ShortnerLamda', {
            runtime: LamdaRuntime.NODEJS_12_X,
            handler: 'index.handler',
            code: LamdaCode.fromAsset('./lambda'),
            environment: {
                'S3_BUCKET': this.bucket.bucketName,
                'S3_PREFIX': 'u',
                'S3_REGION': Stack.of(this).region
            }
        });

        this.bucket.grantDelete(shortLamda);
        this.bucket.grantPut(shortLamda);
        this.bucket.grantRead(shortLamda);

        const api = new RestApi(this, 'ShortnerApi', {
            restApiName: 'Lambda Shortener Service',
            description: 'Rest API for URL Shortener',
        });

        const integration = new LambdaIntegration(shortLamda, {
            proxy: false,
            passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH,
            integrationResponses: [
                {
                    statusCode: '200',
                }
            ],
        });

        api.root.addMethod('POST', integration, {
            methodResponses: [
                {
                    statusCode: '200'
                }
            ]
        });

        const cloudFront = new CloudFrontWebDistribution(this, 'CloudFront', {
            comment: 'CloudFront distribution used as a front end to the server-less URL Shortener',
            originConfigs: [
                {
                    originPath: '',
                    customOriginSource: {
                        domainName: this.bucket.bucketWebsiteDomainName,
                        originProtocolPolicy: OriginProtocolPolicy.HTTP_ONLY
                    },
                    behaviors: [
                        {
                            pathPattern: '/admin/*'
                        },
                        {
                            pathPattern: '/admin'
                        },
                        {
                            pathPattern: '/p/*'
                        },
                        {
                            pathPattern: '/error-pages/*'
                        }, {
                            pathPattern: '/index.html'
                        }, {
                            pathPattern: '/error.html'
                        },
                    ]
                },
                {
                    originPath: '',
                    customOriginSource: {
                        //TODO: This is a hack find a way to fix
                        domainName: `${api.restApiId}.execute-api.${config.region}.${config.urlSuffix}`,
                        originProtocolPolicy: OriginProtocolPolicy.MATCH_VIEWER,
                    },
                    behaviors: [
                        {
                            pathPattern: '/prod/*',
                            allowedMethods: CloudFrontAllowedMethods.ALL,
                            cachedMethods: CloudFrontAllowedCachedMethods.GET_HEAD_OPTIONS,

                        }
                    ]
                },
                {
                    originPath: '/u',
                    customOriginSource: {
                        domainName: this.bucket.bucketWebsiteDomainName,
                        originProtocolPolicy: OriginProtocolPolicy.HTTP_ONLY
                    },
                    behaviors: [
                        {
                            isDefaultBehavior: true
                        }
                    ]
                },
            ],
            errorConfigurations: [
                {
                    errorCode: 404,
                    responseCode: 404,
                    responsePagePath: '/error-pages/error-404.html'
                }
            ],
            aliasConfiguration: {
                acmCertRef: config.certificate.certificateArn,
                names: [config.zoneName],
                sslMethod: SSLMethod.SNI,
                securityPolicy: SecurityPolicyProtocol.TLS_V1_2_2019
            }
        });

        const bucketDeployRoot = new BucketDeployment(this, 'DeployRootFiles', {
            sources: [
                Source.asset('./assets', {
                    exclude: ['p/**', '.DS_Store']
                }),
            ],
            destinationBucket: this.bucket,
            distribution: cloudFront,
            distributionPaths: ['/*']
        });

        this.createPermanentRedirectLink('me', this.meRedirect, cloudFront);

        const record = new ARecord(this, 'ARecord', {
            recordName: config.zoneName,
            zone: config.zone,
            target: RecordTarget.fromAlias(new CloudFrontTarget(cloudFront))
        });

        new CfnOutput(this, 'ShortUrlDns', { value: `https://${record.domainName}` });
    }
}
