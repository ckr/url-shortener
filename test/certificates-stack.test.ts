import { expect, matchTemplate, MatchStyle, haveResource } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { CertificateManager } from '../lib/url-shortner-cert-stack';
import { Config } from '../lib/config';

class TestApp {
    public readonly stack: cdk.Stack;
    public readonly app: cdk.App;

    constructor() {
        this.app = new cdk.App({
            treeMetadata: false
        });
        this.app.node.setContext('stack_params', {
            'dnsZoneName': 'example.com',
            'subjectAlternativeNames': [
                '*.example.com'
            ]
        })
        this.stack = new cdk.Stack(this.app, 'CertificateManagerTestStack', {
            env: {
                account: "1",
                region: 'eu-west-1'
            }
        });
    }
}

describe('Test certificate manager setup', function () {
    test('should create custom resource with correct domain, alternative names and region', () => {
        const app = new TestApp();
        const config = new Config(app.app);
        new CertificateManager(app.stack, 'CertManagerSetup', config);
        expect(app.stack).to(haveResource('AWS::CloudFormation::CustomResource', {
            "DomainName": 'example.com',
            "SubjectAlternativeNames": [
                "*.example.com"
            ],
            "Region": "us-east-1"
        }));
    })
});
