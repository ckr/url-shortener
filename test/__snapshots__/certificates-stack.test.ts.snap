// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test certificate manager setup should create a Certificate Manager Stack based on the saved snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4ArtifactHash5E28809B": Object {
      "Description": "Artifact hash for asset \\"7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4\\"",
      "Type": "String",
    },
    "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3Bucket5BEBDCBE": Object {
      "Description": "S3 bucket for asset \\"7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4\\"",
      "Type": "String",
    },
    "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3VersionKey326451BC": Object {
      "Description": "S3 key for asset version \\"7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CertManagerSetupCertificateCertificateRequestorFunction506E8E55": Object {
      "DependsOn": Array [
        "CertManagerSetupCertificateCertificateRequestorFunctionServiceRoleDefaultPolicy6775976D",
        "CertManagerSetupCertificateCertificateRequestorFunctionServiceRoleB5A3F1F1",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3Bucket5BEBDCBE",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3VersionKey326451BC",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters7af6295e521fd55af94332393ceffb3e866aac4dc4956321f7918f21e72199e4S3VersionKey326451BC",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.certificateRequestHandler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CertManagerSetupCertificateCertificateRequestorFunctionServiceRoleB5A3F1F1",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CertManagerSetupCertificateCertificateRequestorFunctionServiceRoleB5A3F1F1": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CertManagerSetupCertificateCertificateRequestorFunctionServiceRoleDefaultPolicy6775976D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "acm:RequestCertificate",
                "acm:DescribeCertificate",
                "acm:DeleteCertificate",
                "acm:AddTagsToCertificate",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "route53:GetChange",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "route53:changeResourceRecordSets",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":route53:::hostedzone/DUMMY",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CertManagerSetupCertificateCertificateRequestorFunctionServiceRoleDefaultPolicy6775976D",
        "Roles": Array [
          Object {
            "Ref": "CertManagerSetupCertificateCertificateRequestorFunctionServiceRoleB5A3F1F1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CertManagerSetupCertificateCertificateRequestorResource3C9CAE4B": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DomainName": "example.com",
        "HostedZoneId": "DUMMY",
        "Region": "us-east-1",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CertManagerSetupCertificateCertificateRequestorFunction506E8E55",
            "Arn",
          ],
        },
        "SubjectAlternativeNames": Array [
          "*.example.com",
        ],
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
